File: Solutions/977-Squares of a Sorted Array/977-Squares of a Sorted Array.rb
#Third fastest
# @param {Integer[]} a
# @return {Integer[]}
def sorted_squares(a)
    # an array of integers A sorted in non-decreasing order
    #return an array of the squares of each number, also in sorted non-decreasing order.
    a.map{|e| e*e}.sort
end

#Second fastest
# @param {Integer[]} a
# @return {Integer[]}
def sorted_squares(a)
  n = a.size; i = 0
  while i < n && a[i] < 0
    i += 1
  end
  a.map! { _1 * _1 }
  # return a if i == 0
  # return a.reverse if i == n
  j = i - 1
  b = [0] * a.size; k = 0
  while j >= 0 && i < n
    if a[i] < a[j]
      b[k] = a[i]
      i += 1
    else
      b[k] = a[j]
      j -= 1
    end
    k += 1
  end
  while j >= 0
    b[k] = a[j]
    j -= 1
    k += 1
  end
  while i < n
    b[k] = a[i]
    i += 1
    k += 1
  end
  b
end

#Fastest
# @param {Integer[]} a
# @return {Integer[]}
def sorted_squares(a)
    a.map{|x| x*x }.sort
end

File: Solutions/977-Squares of a Sorted Array/977-Squares of a Sorted Array.rs
//Second fastest
//https://leetcode.com/problems/squares-of-a-sorted-array/discuss/330513/Rust-8ms-100
impl Solution {
    pub fn sorted_squares(a: Vec<i32>) -> Vec<i32> {
        let mut left = 0;
        let mut right = a.len() - 1;
        
        let mut ret = vec![];
        while left < right {
            let l = a[left];
            let r = a[right];
            if l.abs() < r.abs() {
                ret.push(r * r);
                right -= 1;
            } else {
                ret.push(l * l);
                left += 1;
            }
        }
        ret.push(a[left] * a[right]); // should point to the same element
        ret.into_iter().rev().collect()
    }
}

//Another
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut res: Vec<i32> = nums.into_iter().map(|a| a.pow(2)).collect();
        res.sort();
        res
    }
}

//Fastest
impl Solution {
    pub fn sorted_squares(a: Vec<i32>) -> Vec<i32> {
        let mut b= Vec::new();
        for n in a{
            b.push(n*n)
        }
        b.sort();
        return b;
    }
}